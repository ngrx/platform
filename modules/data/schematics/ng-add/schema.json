{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxData",
  "title": "NgRx Data Schema",
  "type": "object",
  "properties": {
    "effect": {
      "type": "boolean",
      "default": true,
      "description":
        "When false use the EntityDataModuleWithoutEffects module instead of EntityDataModule."
    },
    "skipPackageJson": {
      "type": "boolean",
      "default": false,
      "description":
        "Do not add @ngrx/data as dependency to package.json (e.g., --skipPackageJson)."
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to the module.",
      "visible": false
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "aliases": ["p"]
    },
    "module": {
      "type": "string",
      "default": "app",
      "description": "Allows specification of the declaring module.",
      "alias": "m",
      "subtype": "filepath"
    },
    "migrateNgrxData": {
      "type": "boolean",
      "default": false,
      "description": "Migrate from ngrx-data, will rename modules.",
      "alias": "migrate"
    },
    "entityConfig": {
      "type": "boolean",
      "default": true,
      "description": "Create the Entity config file"
    }
  },
  "required": []
}
