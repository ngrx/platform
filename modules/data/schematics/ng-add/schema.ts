export interface Schema {
  path?: string;
  effects?: boolean;
  skipPackageJson?: boolean;
  project?: string;
  module?: string;
  migrateNgrxData?: boolean;
  entityConfig?: boolean;
}
