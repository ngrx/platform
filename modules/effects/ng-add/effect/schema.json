{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxEffect",
  "title": "NgRx Effect Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the effect.",
      "type": "string"
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "spec": {
      "type": "boolean",
      "default": true,
      "description": "Specifies if a spec file is generated."
    },
    "module": {
      "type": "string",
      "default": "",
      "description": "Allows specification of the declaring module.",
      "alias": "m",
      "subtype": "filepath"
    },
    "root": {
      "type": "boolean",
      "default": false,
      "description":
        "Flag to indicate whether the effects are registered as root."
    },
    "feature": {
      "type": "boolean",
      "default": false,
      "description": "Flag to indicate if part of a feature schematic."
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description": "Group effects file within 'effects' folder",
      "aliases": ["g"]
    }
  },
  "required": ["name"]
}
