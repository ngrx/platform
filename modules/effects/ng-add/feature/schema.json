{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxFeature",
  "title": "NgRx Feature State Options Schema",
  "type": "object",
  "properties": {
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "name": {
      "description": "The name of the feature.",
      "type": "string"
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "module": {
      "type": "string",
      "description": "Specifies the declaring module.",
      "aliases": ["m"]
    },
    "spec": {
      "type": "boolean",
      "default": true,
      "description": "Specifies if a spec file is generated."
    },
    "reducers": {
      "type": "string",
      "description": "Specifies the reducers file.",
      "aliases": ["r"]
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description":
        "Group actions, reducers and effects within relative subfolders",
      "aliases": ["g"]
    }
  },
  "required": ["name"]
}
