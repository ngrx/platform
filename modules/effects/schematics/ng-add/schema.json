{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxRootEffect",
  "title": "NgRx Root Effect Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the effect.",
      "type": "string",
      "default": "App"
    },
    "skipPackageJson": {
      "type": "boolean",
      "default": false,
      "description":
        "Do not add @ngrx/effects as dependency to package.json (e.g., --skipPackageJson)."
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the effect.",
      "visible": false
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "spec": {
      "type": "boolean",
      "default": true,
      "description": "Specifies if a spec file is generated."
    },
    "module": {
      "type": "string",
      "default": "app",
      "description": "Allows specification of the declaring module.",
      "alias": "m",
      "subtype": "filepath"
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description": "Group effects file within 'effects' folder",
      "aliases": ["g"]
    }
  },
  "required": []
}
