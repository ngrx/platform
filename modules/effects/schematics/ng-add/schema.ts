export interface Schema {
  name: string;
  skipPackageJson?: boolean;
  path?: string;
  project?: string;
  flat?: boolean;
  spec?: boolean;
  module?: string;
  group?: boolean;
}
