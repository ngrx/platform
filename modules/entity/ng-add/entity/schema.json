{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxEntity",
  "title": "NgRx Entity Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the entity.",
      "type": "string"
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "spec": {
      "type": "boolean",
      "description": "Specifies if a spec file is generated.",
      "default": true
    },
    "reducers": {
      "type": "string",
      "description": "Specifies the reducers file.",
      "aliases": ["r"]
    },
    "module": {
      "type": "string",
      "description": "Specifies the declaring module.",
      "aliases": ["m"]
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description":
        "Group actions, reducers and effects within relative subfolders",
      "aliases": ["g"]
    }
  },
  "required": ["name"]
}
