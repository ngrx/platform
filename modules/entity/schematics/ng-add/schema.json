{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxEntity",
  "title": "NgRx Entity Schema",
  "type": "object",
  "properties": {
    "skipPackageJson": {
      "type": "boolean",
      "default": false,
      "description":
        "Do not add @ngrx/entity as dependency to package.json (e.g., --skipPackageJson)."
    }
  },
  "required": []
}
