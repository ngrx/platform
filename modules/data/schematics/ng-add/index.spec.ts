import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import * as path from 'path';
import { Schema as DataEntityOptions } from './schema';
import {
  createWorkspace,
  getTestProjectPath,
} from '@ngrx/schematics-core/testing';

describe('Data ng-add Schematic', () => {
  const schematicRunner = new SchematicTestRunner(
    '@ngrx/data',
    path.join(__dirname, '../collection.json')
  );
  const defaultOptions: DataEntityOptions = {
    skipPackageJson: false,
    project: 'bar',
    module: 'app',
  };

  const projectPath = getTestProjectPath();

  let appTree: UnitTestTree;

  beforeEach(async () => {
    appTree = await createWorkspace(schematicRunner, appTree);
  });

  it('should update package.json', async () => {
    const options = { ...defaultOptions };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const packageJson = JSON.parse(tree.readContent('/package.json'));

    expect(packageJson.dependencies['@ngrx/data']).toBeDefined();
  });

  it('should skip package.json update', async () => {
    const options = { ...defaultOptions, skipPackageJson: true };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const packageJson = JSON.parse(tree.readContent('/package.json'));

    expect(packageJson.dependencies['@ngrx/data']).toBeUndefined();
  });

  it('should import into a specified module', async () => {
    const options = { ...defaultOptions, module: 'app.module.ts' };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(/import { EntityDataModule } from '@ngrx\/data'/);
  });

  it('should fail if specified module does not exist', async () => {
    const options = {
      ...defaultOptions,
      module: `${projectPath}/src/app/app.moduleXXX.ts`,
    };
    let thrownError: Error | null = null;
    try {
      await schematicRunner.runSchematicAsync('data', options, appTree);
    } catch (err: any) {
      thrownError = err;
    }
    expect(thrownError).toBeDefined();
  });

  it('should add entity-metadata config to EntityDataModule', async () => {
    const options = { ...defaultOptions, effects: false, entityConfig: true };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(
      /import { entityConfig } from '.\/entity-metadata'/
    );
    expect(content).toMatch(
      /EntityDataModuleWithoutEffects.forRoot\(entityConfig\)/
    );
  });

  it('should add entity-metadata config file', async () => {
    const options = { ...defaultOptions, entityConfig: true };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    expect(
      tree.files.indexOf(`${projectPath}/src/app/entity-metadata.ts`)
    ).toBeGreaterThanOrEqual(0);
  });

  it('should add entity-metadata config to EntityDataModule', async () => {
    const options = { ...defaultOptions, entityConfig: true };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(
      /import { entityConfig } from '.\/entity-metadata'/
    );
    expect(content).toMatch(/EntityDataModule.forRoot\(entityConfig\)/);
  });

  it('should import EntityDataModuleWithoutEffects into a specified module', async () => {
    const options = {
      ...defaultOptions,
      module: 'app.module.ts',
      effects: false,
      entityConfig: false,
    };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(
      /import { EntityDataModuleWithoutEffects } from '@ngrx\/data'/
    );
  });

  it('should register EntityDataModule in the provided module', async () => {
    const options = { ...defaultOptions, entityConfig: false };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(/EntityDataModule\n/);
  });

  it('should register EntityDataModuleWithoutEffects in the provided module', async () => {
    const options = { ...defaultOptions, effects: false, entityConfig: false };

    const tree = await schematicRunner
      .runSchematicAsync('ng-add', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(/EntityDataModuleWithoutEffects\n/);
  });

  describe('Migration of ngrx-data', () => {
    it('should remove ngrx-data from package.json', async () => {
      const options = { ...defaultOptions, migrateNgrxData: true };

      const packageJsonBefore = JSON.parse(
        appTree.readContent('/package.json')
      );
      packageJsonBefore['dependencies']['ngrx-data'] = '1.0.0';
      appTree.overwrite(
        '/package.json',
        JSON.stringify(packageJsonBefore, null, 2)
      );

      expect(
        JSON.parse(appTree.readContent('/package.json'))['dependencies'][
          'ngrx-data'
        ]
      ).toBeDefined();

      const tree = await schematicRunner
        .runSchematicAsync('ng-add', options, appTree)
        .toPromise();
      const packageJson = JSON.parse(tree.readContent('/package.json'));

      expect(packageJson.dependencies['ngrx-data']).not.toBeDefined();
    });

    it('should rename NgrxDataModule', async () => {
      const options = {
        ...defaultOptions,
        migrateNgrxData: true,
      };

      const dataModulePath = '/data.module.ts';

      const input = `
        import {
          DefaultDataServiceConfig,
          EntityDataService,
          EntityHttpResourceUrls,
          EntityServices,
          Logger,
          NgrxDataModule,
          Pluralizer
        } from 'ngrx-data';

        @NgModule({
          imports: [
            NgrxDataModule.forRoot({
              entityMetadata: entityMetadata,
              pluralNames: pluralNames
            })
          ],
        })
        export class AppModule {}
      `;

      const output = `
        import {
          DefaultDataServiceConfig,
          EntityDataService,
          EntityHttpResourceUrls,
          EntityServices,
          Logger,
          EntityDataModule,
          Pluralizer
        } from '@ngrx/data';

        @NgModule({
          imports: [
            EntityDataModule.forRoot({
              entityMetadata: entityMetadata,
              pluralNames: pluralNames
            })
          ],
        })
        export class AppModule {}
      `;
      appTree.create(dataModulePath, input);

      const tree = await schematicRunner
        .runSchematicAsync('ng-add', options, appTree)
        .toPromise();
      const actual = tree.readContent(dataModulePath);

      expect(actual).toBe(output);
    });

    it('should rename NgrxDataModuleWithoutEffects ', async () => {
      const options = {
        ...defaultOptions,
        migrateNgrxData: true,
      };

      const dataModulePath = '/data.module.ts';

      const input = `
        import {
          DefaultDataServiceConfig,
          EntityDataService,
          EntityHttpResourceUrls,
          EntityServices,
          Logger,
          NgrxDataModuleWithoutEffects,
          Pluralizer
        } from 'ngrx-data';

        @NgModule({
          imports: [
            NgrxDataModuleWithoutEffects.forRoot({
              entityMetadata: entityMetadata,
              pluralNames: pluralNames
            })
          ],
        })
        export class AppModule {}
      `;

      const output = `
        import {
          DefaultDataServiceConfig,
          EntityDataService,
          EntityHttpResourceUrls,
          EntityServices,
          Logger,
          EntityDataModuleWithoutEffects,
          Pluralizer
        } from '@ngrx/data';

        @NgModule({
          imports: [
            EntityDataModuleWithoutEffects.forRoot({
              entityMetadata: entityMetadata,
              pluralNames: pluralNames
            })
          ],
        })
        export class AppModule {}
      `;
      appTree.create(dataModulePath, input);

      const tree = await schematicRunner
        .runSchematicAsync('ng-add', options, appTree)
        .toPromise();
      const actual = tree.readContent(dataModulePath);

      expect(actual).toBe(output);
    });

    it('should rename NgrxDataModuleConfig ', async () => {
      const options = {
        ...defaultOptions,
        migrateNgrxData: true,
      };

      const dataModulePath = '/data.module.ts';

      const input = `
        import {
          DefaultDataServiceConfig,
          EntityDataService,
          EntityHttpResourceUrls,
          EntityServices,
          Logger,
          NgrxDataModule,
          NgrxDataModuleConfig,
          Pluralizer
        } from 'ngrx-data';

        const customConfig: NgrxDataModuleConfig = {
          root: 'api', // default root path to the server's web api
          timeout: 3000, // request timeout
        };

        @NgModule({
          imports: [
            NgrxDataModule.forRoot({
              entityMetadata: entityMetadata,
              pluralNames: pluralNames
            })
          ],
          providers: [
            { provide: DefaultDataServiceConfig, useValue: customConfig },
          ]
        })
        export class AppModule {}
      `;

      const output = `
        import {
          DefaultDataServiceConfig,
          EntityDataService,
          EntityHttpResourceUrls,
          EntityServices,
          Logger,
          EntityDataModule,
          EntityDataModuleConfig,
          Pluralizer
        } from '@ngrx/data';

        const customConfig: EntityDataModuleConfig = {
          root: 'api', // default root path to the server's web api
          timeout: 3000, // request timeout
        };

        @NgModule({
          imports: [
            EntityDataModule.forRoot({
              entityMetadata: entityMetadata,
              pluralNames: pluralNames
            })
          ],
          providers: [
            { provide: DefaultDataServiceConfig, useValue: customConfig },
          ]
        })
        export class AppModule {}
      `;
      appTree.create(dataModulePath, input);

      const tree = await schematicRunner
        .runSchematicAsync('ng-add', options, appTree)
        .toPromise();
      const actual = tree.readContent(dataModulePath);

      expect(actual).toBe(output);
    });
  });
});
