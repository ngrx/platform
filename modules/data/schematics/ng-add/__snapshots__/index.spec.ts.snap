// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data ng-add Schematic Migration of ngrx-data Data ng-add Schematic for standalone application provides data without effects 1`] = `
"import { ApplicationConfig } from '@angular/core';
import { provideEntityData } from '@ngrx/data';
import { entityConfig } from './entity-metadata';

export const appConfig: ApplicationConfig = {
  providers: [provideEntityData(entityConfig)]
};
"
`;

exports[`Data ng-add Schematic Migration of ngrx-data Data ng-add Schematic for standalone application provides data without entityConfig 1`] = `
"import { ApplicationConfig } from '@angular/core';
import { provideEntityData, withEffects } from '@ngrx/data';

export const appConfig: ApplicationConfig = {
  providers: [provideEntityData({}, withEffects())]
};
"
`;

exports[`Data ng-add Schematic Migration of ngrx-data Data ng-add Schematic for standalone application provides default data setup 1`] = `
"import { ApplicationConfig } from '@angular/core';
import { provideEntityData, withEffects } from '@ngrx/data';
import { entityConfig } from './entity-metadata';

export const appConfig: ApplicationConfig = {
  providers: [provideEntityData(entityConfig, withEffects())]
};
"
`;
